File: Source/UI/Frames/DiscordFrame.cs
using WCSharp.Api;

public static class DiscordFrame
{
    private static framehandle EditBox;
    private static framehandle DiscordIconFront;
    private static framehandle DiscordIconBack;
    private static framehandle DiscordText;
    private static framehandle Backdrop;
    private static trigger Trigger;
    private static trigger ESCTrigger;
    private static string Link = "https://discord.gg/GSu6zkNvx5";
    private static string JoinDiscord = "|cffFF2020Join our discord! Highlight then Ctrl + C to copy, Ctrl + V to paste!|r";

    public static void Initialize()
    {
        //BlzLoadTOCFile("war3mapImported\\templates.toc");
        SetupBackdrop();
        CreateFrame();
        RegisterTrigger();
        SetupDiscordIcon();
        ApplyTextFrame();
        RegisterESCTrigger();
    }

    private static void CreateFrame()
    {
        EditBox = framehandle.Create("EscMenuEditBoxTemplate", originframetype.GameUI.GetOriginFrame(0), 0, 0);
        EditBox.SetAbsPoint(framepointtype.Center, 0.4f, 0.165f);
        EditBox.SetSize(0.20f, 0.03f);
        EditBox.TextSizeLimit = Link.Length;
        EditBox.Text = Link;
        EditBox.SetTextAlignment(textaligntype.Center, textaligntype.Center);
    }

    private static void SetupBackdrop()
    {
        Backdrop = framehandle.Create("BACKDROP", "DiscordBackDrop", originframetype.GameUI.GetOriginFrame(0), "QuestButtonDisabledBackdropTemplate", 0);
        Backdrop.SetAbsPoint(framepointtype.Center, 0.4f, 0.21f);
        Backdrop.SetSize(0.195f, 0.035f);
    }

    private static void SetupDiscordIcon()
    {
        DiscordIconFront = framehandle.Create("BACKDROP", "FrameDiscordIconFront", originframetype.GameUI.GetOriginFrame(0), "", 1);
        DiscordIconFront.SetAbsPoint(framepointtype.Center, 0.29f, 0.165f);
        DiscordIconFront.SetSize(0.03f, 0.03f);
        DiscordIconFront.SetTexture("war3mapImported\\DiscordIcon.dds", 0, true);

        DiscordIconBack = framehandle.Create("BACKDROP", "FrameDiscordIconBack", originframetype.GameUI.GetOriginFrame(0), "", 1);
        DiscordIconBack.SetAbsPoint(framepointtype.Center, 0.51f, 0.165f);
        DiscordIconBack.SetSize(0.03f, 0.03f);
        DiscordIconBack.SetTexture("war3mapImported\\DiscordIcon.dds", 0, true);
    }

    private static void ApplyTextFrame()
    {
        DiscordText = framehandle.Create("TEXT", "DiscordText", originframetype.GameUI.GetOriginFrame(0), "", 0);
        DiscordText.SetAbsPoint(framepointtype.Center, 0.4f, 0.21f);
        DiscordText.SetSize(0.19f, 0.03f);
        DiscordText.Text = JoinDiscord;
        DiscordText.SetTextAlignment(textaligntype.Center, textaligntype.Center);
    }

    private static void RegisterTrigger()
    {
        Trigger = trigger.Create();
        Trigger.RegisterFrameEvent(EditBox, frameeventtype.EditBoxEnter);
        Trigger.AddAction(UpdateTextBox);
    }

    private static void RegisterESCTrigger()
    {
        ESCTrigger = trigger.Create();
        foreach (var player in Globals.ALL_PLAYERS)
        {
            ESCTrigger.RegisterPlayerEvent(player, playerevent.EndCinematic);
        }
        ESCTrigger.AddAction(ESCPressed);
    }

    private static void ESCPressed()
    {
        var player = @event.Player;
        if (!player.IsLocal) return;
        Backdrop.Visible = !Backdrop.Visible;
        EditBox.Visible = !EditBox.Visible;
        DiscordIconFront.Visible = !DiscordIconFront.Visible;
        DiscordIconBack.Visible = !DiscordIconBack.Visible;
        DiscordText.Visible = !DiscordText.Visible;
    }

    private static void UpdateTextBox()
    {
        if (!@event.Player.IsLocal) return;
        EditBox.Text = Link;
    }
}

File: src/UI/Frames/DiscordFrame.ts
import { Globals } from 'src/Global/Globals'
import { blzCreateFrame, blzCreateFrameByType, getTriggerPlayer } from 'src/Utility/w3tsUtils'
import { Frame, Trigger } from 'w3ts'

export class DiscordFrame {
    private static EditBox: Frame
    private static DiscordIconFront: Frame
    private static DiscordIconBack: Frame
    private static DiscordText: Frame
    private static Backdrop: Frame
    private static triggerHandle: Trigger
    private static ESCTrigger: Trigger
    private static Link: string = 'https://discord.gg/GSu6zkNvx5'
    private static JoinDiscord: string =
        '|cffFF2020Join our discord! Highlight then Ctrl + C to copy, Ctrl + V to paste!|r'

    public static Initialize() {
        //BlzLoadTOCFile("war3mapImported\\templates.toc");
        DiscordFrame.SetupBackdrop()
        DiscordFrame.CreateFrame()
        DiscordFrame.RegisterTrigger()
        DiscordFrame.SetupDiscordIcon()
        DiscordFrame.ApplyTextFrame()
        DiscordFrame.RegisterESCTrigger()
    }

    private static CreateFrame() {
        DiscordFrame.EditBox = blzCreateFrame(
            'EscMenuEditBoxTemplate',
            Frame.fromHandle(BlzGetOriginFrame(ORIGIN_FRAME_GAME_UI, 0))!,
            0,
            0
        )
        DiscordFrame.EditBox.setAbsPoint(FRAMEPOINT_CENTER, 0.4, 0.165)
        DiscordFrame.EditBox.setSize(0.2, 0.03)
        DiscordFrame.EditBox.textSizeLimit = DiscordFrame.Link.length
        DiscordFrame.EditBox.text = DiscordFrame.Link
        BlzFrameSetTextAlignment(DiscordFrame.EditBox.handle, TEXT_JUSTIFY_CENTER, TEXT_JUSTIFY_CENTER)
    }

    private static SetupBackdrop() {
        DiscordFrame.Backdrop = blzCreateFrameByType(
            'BACKDROP',
            'DiscordBackDrop',
            Frame.fromHandle(BlzGetOriginFrame(ORIGIN_FRAME_GAME_UI, 0))!,
            'QuestButtonDisabledBackdropTemplate',
            0
        )
        DiscordFrame.Backdrop.setAbsPoint(FRAMEPOINT_CENTER, 0.4, 0.21)
        DiscordFrame.Backdrop.setSize(0.195, 0.035)
    }

    private static SetupDiscordIcon() {
        DiscordFrame.DiscordIconFront = blzCreateFrameByType(
            'BACKDROP',
            'FrameDiscordIconFront',
            Frame.fromHandle(BlzGetOriginFrame(ORIGIN_FRAME_GAME_UI, 0))!,
            '',
            1
        )
        DiscordFrame.DiscordIconFront.setAbsPoint(FRAMEPOINT_CENTER, 0.29, 0.165)
        DiscordFrame.DiscordIconFront.setSize(0.03, 0.03)
        DiscordFrame.DiscordIconFront.setTexture('war3mapImported\\DiscordIcon.dds', 0, true)

        DiscordFrame.DiscordIconBack = blzCreateFrameByType(
            'BACKDROP',
            'FrameDiscordIconBack',
            Frame.fromHandle(BlzGetOriginFrame(ORIGIN_FRAME_GAME_UI, 0))!,
            '',
            1
        )
        DiscordFrame.DiscordIconBack.setAbsPoint(FRAMEPOINT_CENTER, 0.51, 0.165)
        DiscordFrame.DiscordIconBack.setSize(0.03, 0.03)
        DiscordFrame.DiscordIconBack.setTexture('war3mapImported\\DiscordIcon.dds', 0, true)
    }

    private static ApplyTextFrame() {
        DiscordFrame.DiscordText = blzCreateFrameByType(
            'TEXT',
            'DiscordText',
            Frame.fromHandle(BlzGetOriginFrame(ORIGIN_FRAME_GAME_UI, 0))!,
            '',
            0
        )
        DiscordFrame.DiscordText.setAbsPoint(FRAMEPOINT_CENTER, 0.4, 0.21)
        DiscordFrame.DiscordText.setSize(0.19, 0.03)
        DiscordFrame.DiscordText.text = DiscordFrame.JoinDiscord
        BlzFrameSetTextAlignment(DiscordFrame.DiscordText.handle, TEXT_JUSTIFY_CENTER, TEXT_JUSTIFY_CENTER)
    }

    private static RegisterTrigger() {
        DiscordFrame.triggerHandle = Trigger.create()!
        DiscordFrame.triggerHandle.triggerRegisterFrameEvent(DiscordFrame.EditBox, FRAMEEVENT_EDITBOX_ENTER)
        DiscordFrame.triggerHandle.addAction(() => DiscordFrame.UpdateTextBox())
    }

    private static RegisterESCTrigger() {
        DiscordFrame.ESCTrigger = Trigger.create()!
        for (const player of Globals.ALL_PLAYERS) {
            DiscordFrame.ESCTrigger.registerPlayerEvent(player, EVENT_PLAYER_END_CINEMATIC)
        }
        DiscordFrame.ESCTrigger.addAction(() => DiscordFrame.ESCPressed())
    }

    private static ESCPressed() {
        const player = getTriggerPlayer()
        if (!player.isLocal()) return
        DiscordFrame.Backdrop.visible = !DiscordFrame.Backdrop.visible
        DiscordFrame.EditBox.visible = !DiscordFrame.EditBox.visible
        DiscordFrame.DiscordIconFront.visible = !DiscordFrame.DiscordIconFront.visible
        DiscordFrame.DiscordIconBack.visible = !DiscordFrame.DiscordIconBack.visible
        DiscordFrame.DiscordText.visible = !DiscordFrame.DiscordText.visible
    }

    private static UpdateTextBox() {
        if (!getTriggerPlayer().isLocal()) return
        DiscordFrame.EditBox.text = DiscordFrame.Link
    }
}
